File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        traverse(root, 0);
        return max;
    }

    private int traverse(TreeNode curr, int depth) {
        if (curr == null) {
            return depth;
        } else {
            int leftDepth = traverse(curr.left, depth + 1);
            int rightDepth = traverse(curr.right, depth + 1);

            if (leftDepth + rightDepth - 2*(depth + 1) > max) {
                max = leftDepth + rightDepth - 2*(depth + 1);
            }

            return Math.max(leftDepth, rightDepth);
        }
    }
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxDist = 0
        def dfs(curr):
            nonlocal maxDist
            if not curr:
                return 0

            leftDepth = dfs(curr.left)
            rightDepth = dfs(curr.right)

            maxDist = max(maxDist, leftDepth + rightDepth)

            return 1 + max(leftDepth, rightDepth)

        dfs(root)
        return maxDist